File: site-tools/boost_site/boostbook_parser.py
#!/usr/bin/env python
# Copyright 2007 Rene Rivera
# Copyright 2011 Daniel James
# Distributed under the Boost Software License, Version 1.0.
# (See accompanying file LICENSE_1_0.txt or http://www.boost.org/LICENSE_1_0.txt)

import xml.dom.minidom, time
from email.utils import parsedate_tz

class BoostBookParser:
    def __init__(self, document = None):
        if document:
            self.document = document
        else:
            self.document = xml.dom.minidom.getDOMImplementation().createDocument(None, "body", None)
    
    def parse(self, filename):
        article = xml.dom.minidom.parse(filename)        

        article_node = article.documentElement
        if article_node.nodeName != 'article':
            print "Boostbook file not article:", filename
            return

        brief_xhtml = self.new_fragment(
            *self.x_children(article_node.getElementsByTagName('articlepurpose')[0])
        )

        title_xhtml = self.new_fragment(
            *self.x_children(article_node.getElementsByTagName('title')[0])
        )

        download_item = None
        download_node = article_node.getElementsByTagName('download')
        if download_node:
            download_item = self.get_child(download_node[0]).data

        pub_date = article_node.getAttribute('last-revision').strip()

        if not pub_date or pub_date[0] == '$':
            pub_date = 'In Progress'
            last_modified = time.time()
        else:
            last_modified = parsedate_tz(pub_date)
            last_modified = time.mktime(last_modified[:-1]) - last_modified[-1]

        description_xhtml = self.x(article_node)
        
        return {
            'title_fragment' : title_xhtml,
            'purpose_fragment' : brief_xhtml,
            'description_fragment' : description_xhtml,
            'pub_date' : pub_date,
            'last_modified' : last_modified,
            'download_item' : download_item
        }

    def x(self, node):
        "Call conversion method for node"

        # This used to deal with multiple arguments and kwargs, but I didn't
        # understand what it was doing, and it never used it, so I simplified
        # it to just deal with a single node.

        name = 'x_'+node.nodeName.replace('-','_').replace('#','_')

        if hasattr(self,name):
            return getattr(self,name)(node)
        else:
            assert False, 'Unknown node type %s'%(name)
            return None
    
    def x_children( self, parent, **kwargs ):
        result = []
        for n in parent.childNodes:
            result.append(self.x(n))
        return result
    
    def x_article(self,node):
        description_xhtml = self.new_fragment()
        for body_item in node.childNodes:
            if body_item.nodeName in ['title', 'articleinfo']:
                continue
            if self.get_child(body_item, tag = 'download'):
                continue
            description_xhtml.appendChild(self.x(body_item))

        return description_xhtml
    
    def x__text(self,node):
        return self.document.createTextNode(node.data);
    
    def x_para(self,node):
        return self.new_node('p',
            *self.x_children(node))

    def x_simpara(self,node):
        return self.new_node('div',
            *self.x_children(node))
        
    def x_ulink(self,node):
        return self.new_node('a',
            href=node.getAttribute('url'),
            *self.x_children(node))
    
    def x_section(self,node):
        return self.new_node('div',
            id=node.getAttribute('id'),
            *self.x_children(node))
    
    def x_title(self,node):
        return self.new_node('h3',
            *self.x_children(node))
    
    def x_link(self,node):
        return self.new_node('span',
            klass='link',
            *self.x_children(node))
    
    def x_itemizedlist(self,node):
        return self.new_node('ul',
            *self.x_children(node))
    
    def x_listitem(self,node):
        return self.new_node('li',
            *self.x_children(node))
    
    def x_phrase(self,node):
        return self.new_node('span',
            klass=node.getAttribute('role'),
            *self.x_children(node))
    
    def x_code(self,node):
        return self.new_node('code',
            *self.x_children(node))
    
    def x_literal(self,node):
        return self.new_node('tt',
            *self.x_children(node))

    def x_emphasis(self,node):
        return self.new_node('em',
            *self.x_children(node))

    def x_inlinemediaobject(self,node):
        image = self.get_child(node,'imageobject')
        if image:
            image = self.get_child(image,'imagedata')
            if image:
                image = image.getAttribute('fileref')
        alt = self.get_child(node,'textobject')
        if alt:
            alt = self.get_child(alt,'phrase')
            if alt and alt.getAttribute('role') == 'alt':
                alt = self.get_child(alt).data.strip()
            else:
                alt = None
        if not alt:
            alt = '[]'
        if image:
            return self.new_node('img',
                src=image,
                alt=alt)
        else:
            return None
    
    def get_child( self, root, tag = None, id = None, name = None):
        for n in root.childNodes:
            found = True
            if tag and found:
                found = found and tag == n.nodeName
            if id and found:
                if n.hasAttribute('id'):
                    found = found and n.getAttribute('id') == id
                else:
                    found = found and n.hasAttribute('id') and n.getAttribute('id') == id
            if name and found:
                found = found and n.hasAttribute('name') and n.getAttribute('name') == name
            if found:
                return n
        return None

    def new_fragment( self, *child ):
        result = self.document.createDocumentFragment()
        for c in child:
            if c:
                result.appendChild(c)
        return result
    
    def new_node( self, tag, *child, **kwargs ):
        result = self.document.createElement(tag)
        for k in kwargs.keys():
            if kwargs[k] != '':
                if k == 'id':
                    result.setAttribute('id',kwargs[k])
                elif k == 'klass':
                    result.setAttribute('class',kwargs[k])
                else:
                    result.setAttribute(k,kwargs[k])
        for c in child:
            if c:
                result.appendChild(c)
        return result
    
    def new_text( self, tag, data, **kwargs ):
        result = self.new_node(tag,**kwargs)
        data = data.strip()
        if len(data) > 0:
            result.appendChild(self.document.createTextNode(data))
        return result

File: common/code/boost_book_parser.php
<?php

# Copyright 2007 Rene Rivera
# Copyright 2011, 2015 Daniel James
# Distributed under the Boost Software License, Version 1.0.
# (See accompanying file LICENSE_1_0.txt or https://www.boost.org/LICENSE_1_0.txt)

class BoostBookParser {
    function parse($filename) {
        $parser = xml_parser_create();
        $state = new BoostBookParser_State();
        xml_parser_set_option($parser, XML_OPTION_CASE_FOLDING, 0);
        if (!xml_parse_into_struct($parser, file_get_contents($filename), $state->values, $state->index)) {
            throw new BoostException("Error parsing XML");
        }
        xml_parser_free($parser);

        $article_node = $state->get_node();
        if ($article_node->get_tag() != 'article') {
            echo "Boostbook file is not an article: {$filename}\n";
            return;
        }

        $id = $article_node->get_attribute('id');

        $brief_xhtml = $this->convert_children_to_xhtml($state->get_first_in_document('articlepurpose'));
        $title_xhtml = $this->convert_children_to_xhtml($state->get_first_in_document('title'));

        $notice_xhtml = null;
        $notice_url = null;
        $notice_node = $state->get_first_in_document('notice');
        if ($notice_node) {
            $notice_xhtml = $this->convert_children_to_xhtml($notice_node);
            $notice_url = $notice_node->get_attribute('url');
        }

        $documentation = null;
        $documentation_node = $state->get_first_in_document('documentation');
        if ($documentation_node) {
            $documentation = $documentation_node->get_value();
        }

        $pub_date = $article_node->get_attribute('last-revision');

        if (!$pub_date or $pub_date[0] == '$') {
            $pub_date = null;
        } else {
            $pub_date = new DateTime($pub_date);
        }

        $description_xhtml = $this->x($state);

        return array(
            'id' => $id,
            'title_xhtml' => $title_xhtml,
            'purpose_xhtml' => $brief_xhtml,
            'description_xhtml' => $description_xhtml,
            'notice_url' => $notice_url,
            'notice_xhtml' => $notice_xhtml,
            'pub_date' => $pub_date,
            'documentation' => $documentation,
        );
    }

    function convert_children_to_xhtml($state) {
        return $state ? $this->x_children($state) : '';
    }

    /** Call conversion method for node */
    function x($state) {
        if ($state->get_type() == 'cdata') {
            return htmlspecialchars($state->get_value(), ENT_NOQUOTES);
        }
        assert($state->get_type() == 'open' || $state->get_type() == 'complete');
        $name = 'x_'.preg_replace('@[-#]@', '_', $state->get_tag());

        if (method_exists($this, $name)) {
            return $this->{$name}($state);
        } else {
            throw new BoostException("Unknown node type {$state->get_tag()}");
        }
    }

    function x_children($state) {
        $tag = $state->get_tag();
        switch($state->get_type()) {
        case 'complete':
            return htmlspecialchars($state->get_value(), ENT_NOQUOTES);
        case 'open':
            $result = htmlspecialchars($state->get_value(), ENT_NOQUOTES);
            for(++$state->pos; $state->get_type() != 'close'; ++$state->pos) {
                $result .= $this->x($state);
            }
            if ($state->get_tag() != $tag) {
                throw new BoostException("Parse error");
            }
            return $result;
        default:
            throw new BoostException("Invalid node in convert_children_to_xhtml: {$state->get_type()}");
        }
    }

    function x_article($state) {
        $description_xhtml = htmlspecialchars($state->get_value(), ENT_NOQUOTES);
        switch($state->get_type()) {
        case 'complete':
            return $description_xhtml;
        case 'open':
            break;
        case 'default':
            throw new BoostException("Error parsing article");
        }
        for(++$state->pos; $state->get_type() != 'close'; ++$state->pos) {
            if (in_array($state->get_tag(), array('title', 'articleinfo'))) {
                $this->skip_to_end_of_tag($state);
                continue;
            }
            if ($state->get_type() == 'open') {
                if ($this->skip_node_with_child($state, array(
                    'download', 'download_basename', 'status', 'notice',
                    'documentation', 'final_documentation')))
                {
                    continue;
                }
            }
            $description_xhtml .= $this->x($state);
        }

        return $description_xhtml;
    }

    function x_para($state) {
        return $this->new_node('p', $this->x_children($state));
    }

    function x_simpara($state) {
        return $this->new_node('div', $this->x_children($state));
    }

    function x_ulink($state) {
        $node = $state->get_node();
        return $this->new_node('a', $this->x_children($state),
            array('href' => $node->get_attribute('url')));
    }

    function x_section($state) {
        $node = $state->get_node();
        return $this->new_node('div', $this->x_children($state),
            array('id' => $node->get_attribute('id')));
    }

    function x_title($state) {
        return $this->new_node('h3', $this->x_children($state));
    }

    function x_link($state) {
        return $this->new_node('span', $this->x_children($state),
            array('class' => 'link'));
    }

    function x_orderedlist($state) {
        return $this->new_node('ol', $this->x_children($state));
    }

    function x_itemizedlist($state) {
        return $this->new_node('ul', $this->x_children($state));
    }

    function x_listitem($state) {
        return $this->new_node('li', $this->x_children($state));
    }

    function x_blockquote($state) {
        return $this->new_node('blockquote', $this->x_children($state));
    }

    function x_phrase($state) {
        $node = $state->get_node();
        return $this->new_node('span', $this->x_children($state),
            array('class' => $node->get_attribute('role')));
    }

    function x_code($state) {
        return $this->new_node('code', $this->x_children($state));
    }

    function x_macroname($state) {
        return $this->new_node('code', $this->x_children($state));
    }

    function x_classname($state) {
        return $this->new_node('code', $this->x_children($state));
    }

    function x_programlisting($state) {
        return $this->new_node('pre', $this->x_children($state));
    }

    function x_literal($state) {
        return $this->new_node('tt', $this->x_children($state));
    }

    function x_subscript($state) {
        return $this->new_node('sub', $this->x_children($state));
    }

    function x_superscript($state) {
        return $this->new_node('sup', $this->x_children($state));
    }

    function x_emphasis($state) {
        $node = $state->get_node();
        $role = '';
        $role = strtolower($node->get_attribute('role'));

        $tags = array(
            '' => 'em',
            'bold' => 'strong',
            'strong' => 'strong'
        );

        if (!isset($tags[$role])) {
            echo "Warning: Unknown emphasis role: {$role}\n";
            $role = '';
        }

        return $this->new_node($tags[$role], $this->x_children($state));
    }

    function x_inlinemediaobject($state) {
        $image = $this->get_child_with_tag($state,'imageobject');
        if ($image) {
            $image = $this->get_child_with_tag($image,'imagedata');
            if ($image) {
                $image = $image->get_attribute('fileref');
            }
        }
        $alt = $this->get_child_with_tag($state,'textobject');
        if ($alt) {
            $alt = $this->get_child_with_tag($alt,'phrase');
            if ($alt && $alt->get_attribute('role') == 'alt') {
                $alt = $this->convert_children_to_xhtml($alt);
            } else {
                $alt = null;
            }
        }
        if (!$alt) {
            $alt = '[]';
        }
        $this->skip_to_end_of_tag($state);
        if ($image) {
            return $this->new_node('img', '', array(
                'src' => $image,
                'alt' => $alt));
        } else {
            return '';
        }
    }

    function get_child_with_tag($state, $tag) {
        switch ($state->get_type()) {
        case 'complete':
            return;
        case 'open':
            $depth = 0;
            $state = clone $state;

            while (true) {
                ++$state->pos;
                if ($depth == 0 && $state->get_tag('tag') == $tag &&
                    in_array($state->get_type(), array('open', 'complete')))
                {
                    return $state;
                }

                if ($state->get_type() == 'close') {
                    if ($depth == 0) { return null; }
                    else { --$depth; }
                }
                else if ($state->get_type() == 'open') {
                    ++$depth;
                }
            }
            break;
        default:
            throw new BoostException("Invalid node in skip_to_end_of_tag");
        }
    }

    function skip_to_end_of_tag($state) {
        switch ($state->get_type()) {
        case 'complete':
            return;
        case 'open':
            $depth = 0;
            for(++$state->pos; $depth || $state->get_type() != 'close'; ++$state->pos) {
                switch($state->get_type()) {
                case 'open': ++$depth; break;
                case 'close': --$depth; break;
                }
            }
            return;
        default:
            throw new BoostException("Invalid node in skip_to_end_of_tag");
        }
    }

    function skip_node_with_child($state, $child_tags) {
        switch ($state->get_type()) {
        case 'complete':
            return false;
        case 'open':
            $found = false;
            $new_state = clone $state;
            $depth = 0;
            for(++$new_state->pos; $depth || $new_state->get_type() != 'close'; ++$new_state->pos) {
                switch($new_state->get_type()) {
                case 'open': ++$depth; break;
                case 'close': --$depth; break;
                }

                if ($depth == 0 && in_array($new_state->get_tag(), $child_tags) &&
                    in_array($new_state->get_type(), array('open', 'complete')))
                {
                    $found = true;
                }
            }
            if ($found) {
                $state->pos = $new_state->pos;
            }
            return $found;
        default:
            throw new BoostException("Invalid node in skip_to_end_of_tag");
        }
    }

    function new_node($tag, $content = '', $attributes = array()) {
        $result = "<{$tag}";
        foreach($attributes as $key => $value) {
            $result .= " {$key}=\"".htmlspecialchars($value, ENT_COMPAT)."\"";
        }
        if (is_null($content) || $content == '') {
            $result .= "/>";
        }
        else {
            $result .= ">";
            $result .= $content;
            $result .= "</{$tag}>";
        }
        return $result;
    }
}

class BoostBookParser_State {
    var $values;
    var $index;
    var $pos = 0;

    function get_node() {
        // Pretty cheesy, but shouldn't be too inefficient as it's a
        // shallow clone.
        return clone $this;
    }

    function get_type() {
        return $this->values[$this->pos]['type'];
    }

    function get_tag() {
        return $this->values[$this->pos]['tag'];
    }

    function get_value() {
        return array_key_exists('value', $this->values[$this->pos]) ?
            $this->values[$this->pos]['value'] : '';
    }

    function get_attribute($name) {
        if (array_key_exists('attributes', $this->values[$this->pos]) &&
            array_key_exists($name, $this->values[$this->pos]['attributes']))
        {
            return $this->values[$this->pos]['attributes'][$name];
        }
        else {
            return null;
        }
    }

    function get_first_in_document($tag) {
        if (!array_key_exists($tag, $this->index)) { return null; }
        $x = clone $this;
        $x->pos = $this->index[$tag][0];
        return $x;
    }
}
